File: app/src/androidTest/java/com/luisenricke/simpleroomapp/database/UserQueryTest.java
package com.luisenricke.simpleroomapp.database;

import android.content.Context;

import androidx.room.Room;
import androidx.test.platform.app.InstrumentationRegistry;
import androidx.test.runner.AndroidJUnit4;

import com.luisenricke.simpleroomapp.database.entity.Pet;
import com.luisenricke.simpleroomapp.database.entity.User;
import com.luisenricke.simpleroomapp.database.dao.UserDAO;

import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

@RunWith(AndroidJUnit4.class)
public class UserQueryTest {

    private UserDAO dao;
    private DatabaseRoom database;

    @Before
    public void setup() {
        Context intrumentationContex = InstrumentationRegistry.getInstrumentation().getContext();
        database = Room.inMemoryDatabaseBuilder(intrumentationContex, DatabaseRoom.class).build();
        dao = database.userDAO();
    }

    @After
    public void finish() {
        database.close();
    }

    @Test
    public void insert() {
        long count = 0;

        User rowOne = new User("test@test", "test");

        dao.insert(rowOne);
        count = dao.count();

        Assert.assertEquals(1, count);
    }

    @Test
    public void insertsList() {
        long count = 0;

        User rowOne = new User("test@test", "test");
        User rowTwo = new User("test2@test", "test2");
        User rowThree = new User("test3@test", "test3");

        List<User> list = Arrays.asList(rowOne, rowTwo, rowThree);
        dao.inserts(list);
        count = dao.count();

        Assert.assertEquals(3, count);
    }

    @Test
    public void insertsVarags() {
        long count = 0;

        User rowOne = new User("test@test", "test");
        User rowTwo = new User("test2@test", "test2");
        User rowThree = new User("test3@test", "test3");

        dao.inserts(rowOne, rowTwo, rowThree);
        count = dao.count();

        Assert.assertEquals(3, count);
    }

    @Test
    public void update() {
        int count = 0;
        User rowOne = new User("test@test", "test");

        dao.insert(rowOne);
        User modified = new User(1, "modified", "modified");
        count = dao.update(modified);

        Assert.assertEquals(1, count);
    }

    @Test
    public void updatesList() {
        User rowOne = new User("test@test", "test");
        User rowTwo = new User("test2@test", "test2");
        User rowThree = new User("test3@test", "test3");

        List<User> list = Arrays.asList(rowOne, rowTwo, rowThree);
        dao.inserts(list);

        User rowOneModified = new User(1, "test@testM", "test");
        User rowTwoModified = new User(2, "test2@testM", "test2");
        User rowThreeNotModified = new User(3, "test3@test", "test3");
        User rowFailUpdate = new User(100, "fail", "fail");
        list = Arrays.asList(rowOneModified, rowTwoModified, rowFailUpdate);

        int aux = dao.updates(list);
        System.out.println(aux + "");

        list = Arrays.asList(rowOneModified, rowTwoModified, rowThreeNotModified);

        List<User> usersDB = dao.get();

        Assert.assertArrayEquals(list.toArray(), usersDB.toArray());
    }

    @Test
    public void updatessVarags() {
        User rowOne = new User("test@test", "test");
        User rowTwo = new User("test2@test", "test2");
        User rowThree = new User("test3@test", "test3");

        dao.inserts(rowOne, rowTwo, rowThree);

        User rowOneModified = new User(1, "test@testM", "test");
        User rowTwoNotModified = new User(2, "test2@test", "test2");
        User rowThreeModified = new User(3, "test3@testM", "test3");
        User rowFailUpdate = new User(100, "fail", "fail");

        int aux = dao.updates(rowOneModified, rowThreeModified, rowFailUpdate);
        System.out.println(aux + "");

        List<User> list = Arrays.asList(rowOneModified, rowTwoNotModified, rowThreeModified);

        List<User> usersDB = dao.get();

        Assert.assertArrayEquals(list.toArray(), usersDB.toArray());
    }

    @Test
    public void drop() {
        long count = 0;

        User rowOne = new User("test@test", "test");
        User rowTwo = new User("test2@test", "test2");
        User rowThree = new User("test3@test", "test3");

        dao.inserts(rowOne, rowTwo, rowThree);

        dao.drop();
        count = dao.count();

        Assert.assertEquals(0, count);
    }

    @Test
    public void delete() {
        long count = 0;
        int deletes = 0;

        User rowOne = new User("test@test", "test");
        User rowTwo = new User("test2@test", "test2");
        User rowThree = new User("test3@test", "test3");

        dao.inserts(rowOne, rowTwo, rowThree);

        deletes = dao.delete(new User(2, "test2@test", "test2"));
        count = dao.count();
        System.out.println("Count of deletes: " + deletes);

        Assert.assertEquals(2, count);
    }

    @Test
    public void deletesList() {
        long count = 0;
        int deletes = 0;

        User rowOne = new User("test@test", "test");
        User rowTwo = new User("test2@test", "test2");
        User rowThree = new User("test3@test", "test3");

        dao.inserts(rowOne, rowTwo, rowThree);

        List<User> deleteRows = Arrays.asList(new User(1, "test@test", "test"),
                new User(2, "test2@test", "test2"),
                new User(3, "test3@test", "test3"));
        deletes = dao.deletes(deleteRows);
        count = dao.count();
        System.out.println("Count of deletes: " + deletes);

        Assert.assertEquals(0, count);
    }

    @Test
    public void deletesVarags() {
        long count = 0;
        int deletes = 0;

        User rowOne = new User("test@test", "test");
        User rowTwo = new User("test2@test", "test2");
        User rowThree = new User("test3@test", "test3");

        dao.inserts(rowOne, rowTwo, rowThree);

        deletes = dao.deletes(new User(1, "test@test", "test"),
                new User(2, "test2@test", "test2"),
                new User(100, "random", "random"));
        count = dao.count();
        System.out.println("Count of deletes: " + deletes);

        Assert.assertEquals(1, count);
    }

    @Test
    public void deleteById() {
        long count = 0;

        User rowOne = new User("test@test", "test");
        User rowTwo = new User("test2@test", "test2");
        User rowThree = new User("test3@test", "test3");

        dao.inserts(rowOne, rowTwo, rowThree);

        count = dao.deleteById(1);

        Assert.assertEquals(1, count);
    }

    @Test
    public void get() {
        User rowOne = new User("test@test", "test");
        User rowTwo = new User("test2@test", "test2");
        User rowThree = new User("test3@test", "test3");

        dao.inserts(rowOne, rowTwo, rowThree);
        List<User> list = dao.get();

        Assert.assertEquals(3, list.size());
    }

    @Test
    public void getByIds() {
        User rowOne = new User("test@test", "test");
        User rowTwo = new User("test2@test", "test2");
        User rowThree = new User("test3@test", "test3");

        dao.inserts(rowOne, rowTwo, rowThree);
        List<User> list = dao.getByIds(new long[]{1, 2, 3});
        List<User> check = dao.get();

        Assert.assertArrayEquals(check.toArray(), list.toArray());
    }
}

File: app/src/androidTest/java/com/luisenricke/simpleroomapp/data/UserQueryTest.kt
package com.luisenricke.simpleroomapp.data

import android.content.Context
import androidx.room.Room
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.runner.AndroidJUnit4
import com.luisenricke.simpleroomapp.data.dao.UserDAO
import com.luisenricke.simpleroomapp.data.entity.User
import org.junit.After
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class UserQueryTest {
    private lateinit var dao: UserDAO
    private lateinit var database: AppDatabase

    @Before fun setup() {
        val instrumentation: Context = InstrumentationRegistry.getInstrumentation().context
        database = Room.inMemoryDatabaseBuilder<AppDatabase>(instrumentation, AppDatabase::class.java).build()
        dao = database.user()
    }

    @After fun finish() {
        database.close()
    }

    @Test fun get() {
        val rowOne = User("test@test", "test")
        val rowTwo = User("test2@test", "test2")
        val rowThree = User("test3@test", "test3")
        dao.inserts(rowOne, rowTwo, rowThree)
        val list = dao.get()
        Assert.assertEquals(3, list.size)
    }

    @Test fun drop() {
        val rowOne = User("test@test", "test")
        val rowTwo = User("test2@test", "test2")
        val rowThree = User("test3@test", "test3")
        dao.inserts(rowOne, rowTwo, rowThree)
        dao.drop()
        val count = dao.count()
        Assert.assertEquals(0, count)
    }

    @Test fun getByIds() {
        val rowOne = User("test@test", "test")
        val rowTwo = User("test2@test", "test2")
        val rowThree = User("test3@test", "test3")
        dao.inserts(rowOne, rowTwo, rowThree)
        val list = dao.get(longArrayOf(1, 2, 3))
        val check = dao.get()
        Assert.assertArrayEquals(check.toTypedArray(), list.toTypedArray())
    }

    @Test fun deleteById() {
        val rowOne = User("test@test", "test")
        val rowTwo = User("test2@test", "test2")
        val rowThree = User("test3@test", "test3")
        dao.inserts(rowOne, rowTwo, rowThree)
        val deletes = dao.delete(2)
        val count = dao.count()
        println("deleted: $deletes")
        Assert.assertEquals(2, count)
    }

    @Test fun deletesByIds() {
        val rowOne = User("test@test", "test")
        val rowTwo = User("test2@test", "test2")
        val rowThree = User("test3@test", "test3")
        dao.inserts(rowOne, rowTwo, rowThree)
        val deleteRows = dao.deletes(longArrayOf(1, 2, 3))
        val count = dao.count()
        println("Count of deletes: $deleteRows")
        Assert.assertEquals(0, count)
    }

    @Test fun insert() {
        val rowOne = User("test@test", "test")
        dao.insert(rowOne)
        val count = dao.count()
        Assert.assertEquals(1, count)
    }

    @Test fun insertsList() {
        val rowOne = User("test@test", "test")
        val rowTwo = User("test2@test", "test2")
        val rowThree = User("test3@test", "test3")
        val list: List<User> = listOf(rowOne, rowTwo, rowThree)
        dao.inserts(list)
        val count = dao.count()
        Assert.assertEquals(3, count)
    }

    @Test fun insertsVararg() {
        val rowOne = User("test@test", "test")
        val rowTwo = User("test2@test", "test2")
        val rowThree = User("test3@test", "test3")
        dao.inserts(rowOne, rowTwo, rowThree)
        val count = dao.count()
        Assert.assertEquals(3, count)
    }

    @Test fun update() {
        val rowOne = User("test@test", "test")
        dao.insert(rowOne)
        val modified = User("modified", "modified", 1)
        val count = dao.update(modified)
        Assert.assertEquals(1, count)
    }

    @Test fun updatesList() {
        val rowOne = User("test@test", "test")
        val rowTwo = User("test2@test", "test2")
        val rowThree = User("test3@test", "test3")
        var list: List<User> = listOf(rowOne, rowTwo, rowThree)
        dao.inserts(list)
        val rowOneModified = User("test@testM", "test", 1)
        val rowTwoModified = User("test2@testM", "test2", 2)
        val rowThreeNotModified = User("test3@test", "test3", 3)
        val rowFailUpdate = User("fail", "fail", 100)
        list = listOf(rowOneModified, rowTwoModified, rowFailUpdate)
        val aux = dao.updates(list)
        println(aux.toString() + "")
        list = listOf(rowOneModified, rowTwoModified, rowThreeNotModified)
        val usersDB = dao.get()
        Assert.assertArrayEquals(list.toTypedArray(), usersDB.toTypedArray())
    }

    @Test fun updatesVararg() {
        val rowOne = User("test@test", "test")
        val rowTwo = User("test2@test", "test2")
        val rowThree = User("test3@test", "test3")
        dao.inserts(rowOne, rowTwo, rowThree)
        val rowOneModified = User("test@testM", "test", 1)
        val rowTwoNotModified = User("test2@test", "test2", 2)
        val rowThreeModified = User("test3@testM", "test3", 3)
        val rowFailUpdate = User("fail", "fail", 100)
        val aux = dao.updates(rowOneModified, rowThreeModified, rowFailUpdate)
        println(aux.toString() + "")
        val list: List<User> = listOf(rowOneModified, rowTwoNotModified, rowThreeModified)
        val usersDB = dao.get()
        Assert.assertArrayEquals(list.toTypedArray(), usersDB.toTypedArray())
    }

    @Test fun delete() {
        val rowOne = User("test@test", "test")
        val rowTwo = User("test2@test", "test2")
        val rowThree = User("test3@test", "test3")
        dao.inserts(rowOne, rowTwo, rowThree)
        val deletes = dao.delete(User("test@test", "test", 1))
        val count = dao.count()
        println("deleted: $deletes")
        Assert.assertEquals(2, count)
    }

    @Test fun deletesList() {
        val rowOne = User("test@test", "test")
        val rowTwo = User("test2@test", "test2")
        val rowThree = User("test3@test", "test3")
        dao.inserts(rowOne, rowTwo, rowThree)
        val deleteRows: List<User> = listOf(
            User("test@test", "test", 1),
            User("test2@test", "test2", 2),
            User("test3@test", "test3", 3)
        )
        val deletes = dao.deletes(deleteRows)
        val count = dao.count()
        println("Count of deletes: $deletes")
        Assert.assertEquals(0, count)
    }

    @Test fun deletesVararg() {
        val rowOne = User("test@test", "test")
        val rowTwo = User("test2@test", "test2")
        val rowThree = User("test3@test", "test3")
        dao.inserts(rowOne, rowTwo, rowThree)
        val deletes = dao.deletes(
            User("test@test", "test", 1),
            User("test2@test", "test2", 2),
            User("test3@test", "test3", 3)
        )
        val count = dao.count()
        println("Count of deletes: $deletes")
        Assert.assertEquals(0, count)
    }

    @Test fun countNothing() {
        val count = dao.count()
        Assert.assertEquals(0, count)
    }


    @Test fun getInEmpty() {
        val list = dao.get()
        Assert.assertEquals(0, list.size)
    }

    @Test fun dropInEmpty() {
        dao.drop()
        val count = dao.count()
        Assert.assertEquals(0, count)
    }

    @Test fun getByIdsInEmpty() {
        val list = dao.get(longArrayOf(1, 2, 3))
        Assert.assertEquals(0, list.size)
    }

    @Test fun deleteByIdInEmpty() {
        val delete = dao.delete(2)
        println("deleted: $delete")
        Assert.assertEquals(0, delete)
    }

    @Test fun deleteByIdNotFound() {
        val rowOne = User("test@test", "test")
        val rowTwo = User("test2@test", "test2")
        val rowThree = User("test3@test", "test3")
        dao.inserts(rowOne, rowTwo, rowThree)
        val delete = dao.delete(5)
        println("deleted: $delete")
        Assert.assertEquals(0, delete)
    }

    @Test fun deletesByIdsInEmpty() {
        val deleteRows = dao.deletes(longArrayOf(1, 2, 3))
        println("Count of deletes: $deleteRows")
        Assert.assertEquals(0, deleteRows)
    }

    @Test fun deletesByIdsNotFoundIds() {
        val rowOne = User("test@test", "test")
        val rowTwo = User("test2@test", "test2")
        val rowThree = User("test3@test", "test3")
        dao.inserts(rowOne, rowTwo, rowThree)
        val deleteRows = dao.deletes(longArrayOf(4, 5, 6))
        val count = dao.count()
        println("Count of deletes: $deleteRows")
        Assert.assertEquals(3, count)
    }

    @Test fun deletesListWithDifferentContentAndSameId() {
        val rowOne = User("test@test", "test")
        val rowTwo = User("test2@test", "test2")
        val rowThree = User("test3@test", "test3")
        dao.inserts(rowOne, rowTwo, rowThree)
        val deleteRows: List<User> = listOf(
            User("random", "random", 1),
            User("random", "random", 2),
            User("random", "random", 3)
        )
        val deletes = dao.deletes(deleteRows)
        val count = dao.count()
        println("Count of deletes: $deletes")
        Assert.assertNotEquals(3, count)
    }

    @Test fun deletesListWithSameContentAndNotFoundIds() {
        val rowOne = User("test@test", "test")
        val rowTwo = User("test2@test", "test2")
        val rowThree = User("test3@test", "test3")
        dao.inserts(rowOne, rowTwo, rowThree)
        val deleteRows: List<User> = listOf(
            User("test@test", "test", 4),
            User("test2@test", "test2", 5),
            User("test3@test", "test3", 6)
        )
        val deletes = dao.deletes(deleteRows)
        val count = dao.count()
        println("Count of deletes: $deletes")
        Assert.assertEquals(3, count)
    }
}
